### ISSUE TITLE
Confuse about how the data is loaded.

### ISSUE THREAD

**Issue by tranxuantuyen** [Lightning-AI/litgpt#560]
I'm a bit confuse by how data is loaded [here](https://github.com/Lightning-AI/lit-gpt/blob/main/finetune/lora.py#L199). Why doesn't the repository use PyTorch's DataLoader? It seems like the current method might be slower because it can't take advantage of multithreading.

Also, instead of training for a set number of epochs, the repository fixes `max_iters` to be equal to the size of the training dataset. However, the random data retrieval strategy in [here](https://github.com/Lightning-AI/lit-gpt/blob/main/finetune/lora.py#L278) doesn't guarantee that the model will see all the data in the dataset.

I guess there must be reasons for this implementation. Is there something I'm missing? Many thanks

**Comment by carmocca** [comment]
Text data is cheap to keep in memory, so there's no need to load them with via worker threads with a DataLoader.

And it's assumed that your dataset is large enough that randomly retrieving from it will converge as expected.

So the reason is just for simplicity. What you suggest could also be implemented and be completely valid.

**Comment by rasbt** [comment]
> Text data is cheap to keep in memory, 

Maybe to clarify here: you mean the training data for finetuning, right?

For pretraining datasets, which are much much larger, it may not be possible to load it into memory. 

> Why doesn't the repository use PyTorch's DataLoader? 

For the finetuning datasets it may be more efficient to just keep it in memory. We do use PyTorch `DataLoader` for the larger datasets for pretraining though, e.g.,

https://github.com/Lightning-AI/lit-gpt/blob/88d51fadcacbc7412f4e0333225be0d23febe5f4/lit_gpt/packed_dataset.py#L27

**Comment by carmocca** [comment]
Yes exactly, I was about finetuning. Thank you for pointing out the pretraining difference.

**Comment by rasbt** [comment]
Thanks to @awaelchli , the data loading was recently revamped and improved. I think we can close this now, but please feel free to reopen if you have any follow-up questions or concerns.
